src/call/client.rs

```diff
@@ -0,0 +1,346 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+use std::marker::PhantomData;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+use futures::{Async, AsyncSink, Future, Poll, Sink, StartSend, Stream};
+use grpc_sys;
+use protobuf::{self, Message, MessageStatic};
+
+use async::{BatchMessage, CqFuture};
+use call::Call;
+use error::Error;
+use super::{SinkBase, StreamingBase};
+
+/// Update the flag bit in res.
+#[inline]
+fn change_flag(res: &mut u32, flag: u32, set: bool) {
+    if set {
+        *res |= flag;
+    } else {
+        *res &= !flag;
+    }
+}
+
+#[derive(Default)]
+pub struct CallOption {
+    timeout: Option<Duration>,
+    write_flags: u32,
+    call_flags: u32,
+}
+
+impl CallOption {
+    /// Signal that the call is idempotent
+    pub fn with_idempotent(mut self, is_idempotent: bool) -> CallOption {
```

Comment:
`with_` means constructor, use `idempotent` as a builder chaining function name is ok